sklearn/gaussian_process/kernels.py

```diff
@@ -0,0 +1,475 @@
+""" Kernels for Gaussian process regression and classification.
+
+The kernels in this module allow kernel-engineering, i.e., they can be
+combined via the "+" and "*" operators. These expressions can also contain
+scalar values, which are automatically converted to a constant kernel.
+
+All kernel allow (analytic) gradient-based hyperparameter optimization.
+The space of hyperparameters can be specified by giving lower und upper
+boundaries for the value of each hyperparameter (the search space is thus
+rectangular). This can be achieved by using a pair or triple instead of a
+single float wherever a parameter value is specified. In case of a pair,
+the first value specifies the lower boundary and the second value the upper
+boundary. In case of a triple, the middle value specified the initial value
+of the parameter during hyperparameter-optimization.
```

Comment:
Bounds are useful, e.g., when there are multiple local optima and you want to exclude some of those. For instance the high-noise solution from plot_gpr_noisy.py. Another use-case is to start the optimizer from different random points in the search space to find different local optima and select the best one at the end. If no bounds are specified, it is unclear how to sample random start points.

We could also make the bounds 0 and +inf per default and allow other optimizers from scipy in that case (provided that we log-transform the search space such that it actually becomes [-inf, +inf]).